# Exercise 1: Business cycle correlations

For this exercise, you'll be using macroeconomic data from the folder `data/FRED`.

1.  There are seven decade-specific files named `FRED_monthly_19X0.csv` where `X` identifies the decade (`X` takes on the values 5, 6, 7, 8, 9, 0, 1). Write a loop that reads in all seven files as DataFrames and store them in a list.

    *Hint:* Recall from the lecture that you should use `pd.read_csv(..., parse_dates=['DATE'])` to automatically parse strings stored in the `DATE` column as dates.
2.  Use [`pd.concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) to concate these data sets into a single `DataFrame` and set the `DATE` column as the index.
3.  You realize that your data does not include GDP since this variable is only reported at quarterly frequency.
    Load the GDP data from the file `GDP.csv` and merge it with your monthly data using an _inner join_.
4.  You want to compute how (percent) changes of the variables in your data correlate with percent changes in GDP.

    1. Create a _new_ `DataFrame` which contains the percent changes in CPI and GDP (using 
    [`pct_change()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pct_change.html),
    and the absolute changes for the remaining variables (using 
    [`diff()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.diff.html)).
    2.  Compute the correlation of the percent changes in GDP with the (percent) changes of all other variables using [`corr()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html). What does the sign and magnitude of the correlation coefficient tell you?



In [1]:
import pandas as pd
DATA_PATH = '../../data/FRED'
data = []
for x in range(1950,2011,10):
    filename = f'{DATA_PATH}/FRED_monthly_{x}.csv'
    df = pd.read_csv(filename, parse_dates=['DATE'])
    data.append(df)


In [2]:
df = pd.concat(data,ignore_index=True)
df = df.set_index('DATE')

In [3]:
df_gdp = pd.read_csv('../../data/FRED/GDP.csv', parse_dates=['DATE'])

In [4]:
df = pd.merge(df.reset_index(),df_gdp.reset_index(),how='inner',on='DATE')

In [5]:
df

DATE    CPI  UNRATE  FEDFUNDS  REALRATE  LFPART  index      GDP
0   1950-01-01   23.5     6.5       NaN       NaN    58.9     12   2346.1
1   1950-04-01   23.6     5.8       NaN       NaN    59.2     13   2417.7
2   1950-07-01   24.1     5.0       NaN       NaN    59.1     14   2511.1
3   1950-10-01   24.5     4.2       NaN       NaN    59.4     15   2559.2
4   1951-01-01   25.4     3.7       NaN       NaN    59.1     16   2594.0
..         ...    ...     ...       ...       ...     ...    ...      ...
275 2018-10-01  252.8     3.8       2.2      -0.2    62.9    287  20304.9
276 2019-01-01  252.6     4.0       2.4       0.6    63.1    288  20431.6
277 2019-04-01  255.2     3.7       2.4       3.1    62.8    289  20602.3
278 2019-07-01  255.8     3.7       2.4       1.1    63.1    290  20843.3
279 2019-10-01  257.2     3.6       1.8      -0.0    63.3    291  20985.4

[280 rows x 8 columns]

In [6]:
df = df.set_index('DATE')

In [7]:
df

CPI  UNRATE  FEDFUNDS  REALRATE  LFPART  index      GDP
DATE                                                                 
1950-01-01   23.5     6.5       NaN       NaN    58.9     12   2346.1
1950-04-01   23.6     5.8       NaN       NaN    59.2     13   2417.7
1950-07-01   24.1     5.0       NaN       NaN    59.1     14   2511.1
1950-10-01   24.5     4.2       NaN       NaN    59.4     15   2559.2
1951-01-01   25.4     3.7       NaN       NaN    59.1     16   2594.0
...           ...     ...       ...       ...     ...    ...      ...
2018-10-01  252.8     3.8       2.2      -0.2    62.9    287  20304.9
2019-01-01  252.6     4.0       2.4       0.6    63.1    288  20431.6
2019-04-01  255.2     3.7       2.4       3.1    62.8    289  20602.3
2019-07-01  255.8     3.7       2.4       1.1    63.1    290  20843.3
2019-10-01  257.2     3.6       1.8      -0.0    63.3    291  20985.4

[280 rows x 7 columns]

In [12]:
df.set_index('DATE')

CPI  UNRATE  FEDFUNDS  REALRATE  LFPART  INFLATION
DATE                                                           
1950-01-01  23.5     6.5       NaN       NaN    58.9        NaN
1950-02-01  23.6     6.4       NaN       NaN    58.9        NaN
1950-03-01  23.6     6.3       NaN       NaN    58.8        NaN
1950-04-01  23.6     5.8       NaN       NaN    59.2        NaN
1950-05-01  23.8     5.5       NaN       NaN    59.1        NaN
...          ...     ...       ...       ...     ...        ...
1969-08-01  36.9     3.5       9.2       NaN    60.3        NaN
1969-09-01  37.1     3.7       9.2       NaN    60.3        NaN
1969-10-01  37.3     3.7       9.0       NaN    60.4        NaN
1969-11-01  37.5     3.5       8.8       NaN    60.2        NaN
1969-12-01  37.7     3.5       9.0       NaN    60.2        NaN

[1777 rows x 6 columns]

In [13]:
df_change = df[['GDP','CPI']].pct_change()*100

KeyError: "['GDP'] not in index"

***
# Exercise 2: Loading many data files

In the previous exercise, you loaded the individual files by specifing an explicit list of file names. This can become tedious or infeasible if your data is spread across many files with varying file name patterns. Python offers the possibility to iterate over all files in a directory (for example, using [`os.listdir()`](https://docs.python.org/3/library/os.html#os.listdir)),
or to iterate over files that match a pattern, for example using [`glob.glob()`](https://docs.python.org/3/library/glob.html).

Repeat parts (1) and (2) from the previous exercise, but now iterate over the input files using 
[`glob.glob()`](https://docs.python.org/3/library/glob.html). You'll need to use a wildcard `*` and make sure to match only the relevant files in `data/FRED`, i.e., those that start with `FRED_monthly_1` or `FRED_monthly_2`.

In [14]:
import glob
pattern = f'{DATA_PATH}/FRED_monthly_[12].csv'
data = []
for file in glob.glob(pattern):
    print(f'proccessing file: {file}')
    df = pd.read_csv(file)
    data.append(df)
    

In [15]:
df.sort_values('DATE').reset_index(drop=True)

DATE    CPI  UNRATE  FEDFUNDS  REALRATE  LFPART  INFLATION
0     1947-01-01   21.5     NaN       NaN       NaN     NaN        NaN
1     1947-02-01   21.6     NaN       NaN       NaN     NaN        NaN
2     1947-03-01   22.0     NaN       NaN       NaN     NaN        NaN
3     1947-04-01   22.0     NaN       NaN       NaN     NaN        NaN
4     1947-05-01   22.0     NaN       NaN       NaN     NaN        NaN
...          ...    ...     ...       ...       ...     ...        ...
1772  2024-09-01  314.9     4.1       5.1       2.4    62.7        2.4
1773  2024-10-01  315.6     4.1       4.8       2.2    62.5        2.6
1774  2024-11-01  316.4     4.2       4.6       2.1    62.5        2.7
1775  2024-12-01  317.6     4.1       4.5       1.6    62.5        2.9
1776  2025-01-01    NaN     NaN       NaN       NaN     NaN        3.0

[1777 rows x 7 columns]

In [16]:
df.set_index('DATE').sort_index()

CPI  UNRATE  FEDFUNDS  REALRATE  LFPART  INFLATION
DATE                                                            
1947-01-01   21.5     NaN       NaN       NaN     NaN        NaN
1947-02-01   21.6     NaN       NaN       NaN     NaN        NaN
1947-03-01   22.0     NaN       NaN       NaN     NaN        NaN
1947-04-01   22.0     NaN       NaN       NaN     NaN        NaN
1947-05-01   22.0     NaN       NaN       NaN     NaN        NaN
...           ...     ...       ...       ...     ...        ...
2024-09-01  314.9     4.1       5.1       2.4    62.7        2.4
2024-10-01  315.6     4.1       4.8       2.2    62.5        2.6
2024-11-01  316.4     4.2       4.6       2.1    62.5        2.7
2024-12-01  317.6     4.1       4.5       1.6    62.5        2.9
2025-01-01    NaN     NaN       NaN       NaN     NaN        3.0

[1777 rows x 6 columns]

***
# Exercise 3: Weekly returns of the magnificent seven

In this exercise, you are asked to analyze the weekly stockmarket returns
of the so-called magnificent 7 which are some of the most successful tech companies 
of the last decades years:
Apple (AAPL), Amazon (AMZN), Alphabet/Google (GOOGL), Meta (META), Microsoft (MSFT), Nvidia (NVDA), and Tesla (TSLA).

The data for this exercise is located in the folder `data/stockmarket/`.

1.  For each of the seven stocks listed above, there is a corresponding 
    CSV file in this directory (based on the ticker symbol).

    1.  For each ticker symbol, load the corresponding CSV file and make sure 
        that the `Date` is set as the index.

        The DataFrame has two columns, `Open` and `Close`, which contain the 
        opening and closing price for each trading day.

    3.  Use [`resample()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.resample.html)
        to resample the daily data to a weekly frequency by specifying `resample('W')`,
        and compute the weekly returns in percent:

        $$
        \text{Weekly returns} = \frac{\text{Close price on last day} - \text{Open price on first day}}{\text{Open price on first day}} \times 100
        $$

        *Hint:* You can obtain the first and last observation using the 
        [`first()`](https://pandas.pydata.org/docs/reference/api/pandas.core.resample.Resampler.first.html) and 
        [`last()`](https://pandas.pydata.org/docs/reference/api/pandas.core.resample.Resampler.last.html)
        methods.

    4.  Append these returns to a list so you can merge them into a single DataFrame later.

2.  Merge the list of weekly returns you computed into a single DataFrame.
    Keep only the intersection of dates available for all 7 stocks.

    *Hint:* This can be achieved using either 
    [`pd.concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html),
    [`pd.merge()`](https://pandas.pydata.org/docs/reference/api/pandas.merge.html), or 
    [`DataFrame.join()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html).

3.  Finally, you are interested in how the weekly returns are correlated across 
    the 7 stocks. 

    1.  Compute and report the pairwise correlations using 
        [DataFrame.corr()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html).

    2.  Create a figure with 7-by-7 subplots showing the pairwise scatter plots of weekly returns 
        for each combination of stocks.

        You can do this either with the
        [`scatter_matrix()`](https://pandas.pydata.org/docs/reference/api/pandas.plotting.scatter_matrix.html) function contained in `pandas.plotting`, 
        or manually build the figure using Matplotlib functions.

    3.  **[Advanced]**
        In each of the subplots, add a text that reports the pairwise correlation
        for these stocks which you computed earlier.
        (e.g., the correlation between returns on AAPL and AMZN is about 0.42,
        so this text should be added to the subplot showing the 
        scatter plot of AAPL vs. AMZN).


In [34]:
DATA_PATH = '../../data/stockmarket'

data = []
files = ['/AAPL.csv', '/MSFT.csv', '/GOOGL.csv', '/AMZN.csv', '/TSLA.csv', '/META.csv', '/NVDA.csv']
for file in files:
    df = pd.read_csv(f'{DATA_PATH}{file}',parse_dates=True, index_col='Date')
    data.append(df)



In [36]:
df

Open     Close
Date                          
1999-01-22    0.0401    0.0376
1999-01-25    0.0406    0.0415
1999-01-26    0.0420    0.0383
1999-01-27    0.0384    0.0382
1999-01-28    0.0382    0.0381
...              ...       ...
2024-12-23  136.2503  139.6396
2024-12-24  139.9695  140.1895
2024-12-26  139.6696  139.8995
2024-12-27  138.5198  136.9801
2024-12-30  134.8006  137.4601

[6527 rows x 2 columns]

In [37]:
wk_rt = []
for df in data:
    df.resample('W')
    weekly_ret = (df['Close']-df['Open'])/df['Open']*100
    wk_rt.append(weekly_ret)

In [38]:
wk_rt

[Date
 1980-12-12    0.000000
 1980-12-15   -0.533049
 1980-12-16   -0.460299
 1980-12-17    0.000000
 1980-12-18    0.000000
                 ...   
 2024-12-23    0.196243
 2024-12-24    1.060741
 2024-12-26    0.321444
 2024-12-27   -0.868760
 2024-12-30   -0.011896
 Length: 11104, dtype: float64,
 Date
 1986-03-13    9.778598
 1986-03-14    3.529412
 1986-03-17    1.785714
 1986-03-18   -2.551834
 1986-03-19   -1.800327
                 ...   
 2024-12-23   -0.341174
 2024-12-24    1.076723
 2024-12-26   -0.220917
 2024-12-27   -0.936503
 2024-12-30   -0.288697
 Length: 9778, dtype: float64,
 Date
 2004-08-19    0.337974
 2004-08-20    7.225652
 2004-08-23   -1.217075
 2004-08-24   -5.725757
 2004-08-25    0.992870
                 ...   
 2024-12-23    1.043470
 2024-12-24    0.651831
 2024-12-26    0.230624
 2024-12-27   -1.123373
 2024-12-30    0.758735
 Length: 5126, dtype: float64,
 Date
 1997-05-15   -19.688269
 1997-05-16   -12.093496
 1997-05-19    -2.954545
 1997-05-20    

***
# Exercise 4: Decade averages of macro time series


For this exercise, you'll be using macroeconomic data from the folder `data/FRED`.

1.  There are five files containing monthly observations on annual inflation (INFLATION), the Fed Funds rate (FEDFUNDS), the labor force participation rate (LFPART), the 1-year real interest rate (REALRATE) and the unemployment rate (UNRATE).

    1.  Write a loop to import these files and store the individual DataFrames in a list.

        *Hint:* Recall from the lecture that you should use 
        `pd.read_csv(..., parse_dates=['DATE'], index_col='DATE')` to automatically parse strings stored in the `DATE` column as dates and set the `DATE`
        column as the index.

    2.  Use 
        [`pd.concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html)
        to concatenate this list of DataFrames along the column dimension
        using an outer join (`join='outer'`) to obtain a merged data set.

3.  You want to compute the average value of each variable by decade, but you want to include only decades without _any_ missing values for _all_ variables.

    1.  Create a variable `Decade` which stores the decade (1940, 1950, ...) for each observation.

        *Hint:* You should have set the `DATE` as the `DataFrame` index. Then you can access the calendar year using the attribute `df.index.year` which can be used to compute the decade.

    2.  Create an indicator variable which takes on the value `True` 
        whenever all observations (all columns) for a given date are non-missing, and `False`
        if at least one variable has a missing observation. 

    3.  Aggregate this indicator to decades using a
    [`groupby()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html) so that the indicator takes on the value `True` whenever
    _all_ variables in a given decade have no missing values, and `False`
    otherwise.

        *Hint:* You can use the 
        [`all()`](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.all.html) aggregation for this.

    4.  Merge this decade-level indicator data back into the original `DataFrame` (_many-to-one_ merge). 
4.  Using this indicator, drop all observations which are in a decade with missing values.
5.  Compute the decade average for each variable.

**Challenge**

-   Your pandas guru friend claims that all the steps in 2.2 to 2.4 can be done with a single one-liner using [`transform()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.transform.html). Can you come up with a solution?

    

***
# Exercise 5: Merging additional Titanic data

In this exercise, you'll be working with the the original Titanic data set in `titanic.csv` and additional (partly fictitious) information on passengers stored in `titanic-additional.csv`, both located in the `data/` folder.

The goal of the exercise is to calculate the survival rates by country of residence (for this exercise we restrict ourselves to the UK, so these will be England, Scotland, etc.).

1.  Load the `titanic.csv` and `titanic-additional.csv` into two DataFrames.

    Inspect the columns contained in both data sets. As you can see, the original data contains the full name including the title
    and potentially maiden name (for married women) in a single column.
    The additional data contains this information in separate columns.
    You want to merge these data sets, but you first need to create common keys in both DataFrames.

2.  Since the only common information is the name, you'll need to extract the individual name components from the original DataFrame
    and use these as merge keys.

    Focusing only on men (who have names that are much easier to parse), split the `Name` column into the tokens 
    `Title`, `FirstName` and `LastName`, just like the columns in the second DataFrame.

    *Hint:* This is the same task as in the last exercise in Workshop 2. You can just use your solution here.

3.  Merge the two data sets based on the columns `Title`, `FirstName` and `LastName` you just created using a _left join_ (_one-to-one_ merge).
    Tabulate the columns and the number of non-missing observations to make sure that merging worked. 

    *Note:* The additional data set contains address information only for passengers from the UK, so some of these fields will be missing.

4.  You are now in a position to merge the country of residence (_many-to-one_ merge). Load the country data from `UK_post_codes.csv` which contains 
    the UK post code prefix (which you can ignore), the corresponding city, and the corresponding country.

    Merge this data with your passenger data set using a _left join_ (what is the correct merge key?).

5.  Tabulate the number of observations by `Country`, including the number of observations with missing `Country` (these are passengers residing outside the UK).

    Finally, compute the mean survival rate by country.